### PR TITLE
daemon: Create all global maps in cilium-agent

### LINKED ISSUE #9276: Open or create all maps from cilium-agent
Currently (early v1.7 cycle), some of the global maps used by multiple BPF programs are opened/created from `cilium-agent` in the function below:

https://github.com/cilium/cilium/blob/d0f2435b4c423dbff571ad8bcf206c177a18eef9/daemon/datapath.go#L530-L533

But others (such as conntrack, or the global tail call map) are not.

Note that we must make sure that if any map is being created, its attributes *must* match the attributes defined in the C templates, otherwise when we load a program for an endpoint the first time, it will delete/recreate the map.

Tasks:
* [x] Gather a list of globally shared maps that are not opened/created in the daemon (see `pkg/maps/` and `bpf/`)
* [x] Make sure that the map attributes are guaranteed to be in sync for those map types between the BPF C definitions and the Golang definitions
* [x] Add `OpenOrCreate()` calls for all maps to `initMaps()` so we can be sure that the globally shared maps are guaranteed to be in place before the first BPF program load
  * Some maps, such as the tail call maps, are never used from the `cilium-agent`. These should be created, and pinned to the filesystem, then the fd closed.

### PR DESCRIPTION
**Pull request by pchaigno**
The first 6 commits are preparatory work for the last 9 commits, which creates all global maps in the cilium-agent. The last 9 commits can be easily squashed if preferred as they're fairly similar (I initially broke them down to bisect a bug).

Fixes: #9276

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/10241)

### CODE REVIEW

**Comment by pchaigno** [comment]
test-me-please

File: pkg/maps/ctmap/ctmap.go [hunk 1]
```diff
@@ -249,19 +259,20 @@ func (m *Map) DumpEntries() (string, error) {
 
 // NewMap creates a new CT map of the specified type with the specified name.
 func NewMap(mapName string, mapType MapType) *Map {
```
Comment (tklauser) [inline pkg/maps/ctmap/ctmap.go]: While at it, I think you could unexport this function (maybe in a separate commit). It is only used within the `ctmap` package.

Comment (pchaigno) [reply pkg/maps/ctmap/ctmap.go]: Done.

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

File: cilium/cmd/bpf_ct_flush.go [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2017-2018 Authors of Cilium
+// Copyright 2017-2020 Authors of Cilium
```
Comment (qmonnet) [inline cilium/cmd/bpf_ct_flush.go]: This file is not changed otherwise in the PR, so let's remove copyright year change?

Comment (qmonnet) [reply cilium/cmd/bpf_ct_flush.go]: Not related, but there's a typo in this commit's log: “were previously always create” -> created (but GH doesn't offer to comment on commit log it seems).

File: cilium/cmd/bpf_ct_list.go [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2017-2018 Authors of Cilium
+// Copyright 2017-2020 Authors of Cilium
```
Comment (qmonnet) [inline cilium/cmd/bpf_ct_list.go]: Same here, no change in file.

File: pkg/maps/ctmap/gc/gc.go [hunk 4]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2016-2019 Authors of Cilium
+// Copyright 2016-2020 Authors of Cilium
```
Comment (qmonnet) [inline pkg/maps/ctmap/gc/gc.go]: File not changed.

File: pkg/bpf/map_linux.go [hunk 5]
```diff
@@ -452,6 +452,16 @@ func (m *Map) OpenOrCreateUnpinned() (bool, error) {
...
 }
 
+// Create is similar to OpenOrCreate, but closes the map after creating or
+// opening it.
+func (m *Map) Create() (bool, error) {
+	isNew, err := m.OpenOrCreate()
+	if err != nil {
+		return isNew, err
+	}
+	return isNew, m.Close()
```
Comment (qmonnet) [inline pkg/bpf/map_linux.go:462]: Commit log: “doesn't open” -> “doesn't keep open”?

File: pkg/maps/neighborsmap/neighborsmap.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+		)
+		if _, err := neighbors4Map.Create(); err != nil {
+			return err
+		}
+	}
+	if ipv6 {
+		neighbors6Map := bpf.NewMap(Map6Name,
+			bpf.MapTypeLRUHash,
+			&Key6{},
+			int(unsafe.Sizeof(Key4{})),
```
Comment (qmonnet) [inline pkg/maps/neighborsmap/neighborsmap.go]: Should this be `Key6{}` instead?

**Review by qmonnet — CHANGES_REQUESTED** [review]

**Comment by pchaigno** [comment]
test-me-please

**Pull request closed without merging**